Cryptlib/SysCall/BaseMemAllocation.c

```diff
@@ -19,20 +19,63 @@ WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
 // -- Memory-Allocation Routines --
 //
 
+/* Prefix allocations with the allocation's total size.
+ *
+ * This is used to implement `realloc`, which needs to know the
+ * allocation's current size so that the data can be copied over to the
+ * new allocation.
+ */
+struct AllocWithSize {
+  /* Size of the whole allocation, including this `size` field.
+   *
+   * Allocations returned by AllocatePool are always eight-byte aligned
+   * (according to the UEFI spec), so use an eight-byte `size` field to
+   * make the `data` field also eight-byte aligned. */
+  UINT64 size;
+
+  /* The beginning of the allocation returned to the caller of malloc.
+   *
+   * The pointer to this data is what will then get passed into realloc
+   * and free; those functions use ptr_to_alloc_with_size to get the
+   * pointer to the underlying pool allocation. */
+  UINT8 data[0];
+};
+
+/* Convert a `ptr` (as returned by malloc) to an AllocWithSize by
+ * subtracting eight bytes. */
+static struct AllocWithSize *ptr_to_alloc_with_size (void *ptr) {
+  UINT8 *cptr = (UINT8*) ptr;
+  return (struct AllocWithSize*) (cptr - sizeof(struct AllocWithSize));
+}
+
 /* Allocates memory blocks */
 void *malloc (size_t size)
 {
-  return AllocatePool ((UINTN) size);
+  UINTN alloc_size = (UINTN) (size + sizeof(struct AllocWithSize));
+
+  struct AllocWithSize *alloc = AllocatePool (alloc_size);
+  if (alloc) {
+    alloc->size = alloc_size;
+    return alloc->data;
+  } else {
+    return NULL;
+  }
 }
 
 /* Reallocate memory blocks */
 void *realloc (void *ptr, size_t size)
 {
-  //
-  // BUG: hardcode OldSize == size! We have no any knowledge about
-  // memory size of original pointer ptr.
-  //
-  return ReallocatePool (ptr, (UINTN) size, (UINTN) size);
+  struct AllocWithSize *alloc = ptr_to_alloc_with_size (ptr);
+  UINTN old_size = alloc->size;
+  UINTN new_size = (UINTN) (size + sizeof(struct AllocWithSize));
+
+  alloc = ReallocatePool (alloc, old_size, new_size);

```

Comment:
Two things to consider here.

**(1)** At this point, we're past the `malloc()` shortcut (which I pointed out above, for `ptr==NULL`); that is, we're actually resizing. It's still possible that we're resizing to zero (`size==0`). In that case, the logic here will preserve a just the "empty" `AllocWithSize` structure. That actually does conform to POSIX, again:

https://pubs.opengroup.org/onlinepubs/9699919799/functions/realloc.html

and returning the `alloc->data` flexible array member again conforms to the ISO C99 standard; **but* again this deserves a comment in the code.

**(2)** Unfortunately, the other thing to consider is a nasty preexistent bug, inside ReallocatePool().

As of commit 657b2483ca6e, shim consumes its own fork of gnu-efi (the submodule) at commit 03670e14f263. And at that point, the `ReallocatePool` function in `shim/gnu-efi/lib/misc.c` looks like this:

~~~
VOID *
ReallocatePool (
    IN VOID                 *OldPool,
    IN UINTN                OldSize,
    IN UINTN                NewSize
    )
{
    VOID                    *NewPool;

    NewPool = NULL;
    if (NewSize) {
        NewPool = AllocatePool (NewSize);
    }

    if (OldPool) {
        if (NewPool) {
            CopyMem (NewPool, OldPool, OldSize < NewSize ? OldSize : NewSize);
        }
    
        FreePool (OldPool);
    }
    
    return NewPool;
}
~~~

You do realize that this `ReallocatePool()` implementation frees `OldPool` even in case it attempts to, but *fails*, to allocate `NewPool`, right? That is, even in case the caller does not intend to use `ReallocatePool()` in a `FreePool()` role, but in an actual resizing role. Note that the releasing of `OldPool` *only* depends on `OldPool` being non-NULL on entry.

So, from our `realloc()` function, we're certainly calling `ReallocatePool()` with:
* a nonzero `old_size` (see the assertion I point out above),
* a nonzero `new_size`,
* and a non-NULL "old pool" pointer (because we're going to handle `ptr==NULL` on its own separate path, earlier).

That means this ReallocatePool() execution, internal to our realloc() implementation, is an actual *resizing* operation. And if the new area cannot be allocated, we're going to trash the existent area nevertheless. We're going to propagate `NULL` out to the caller of `realloc()`, and they will think that, while `realloc()` surely failed, they could continue accessing the *old* (original) allocation. They're going to think that because that's what POSIX guarantees -- but this implementation of ReallocatePool() violates.

In other words, `ReallocatePool()` itself is buggy.

... It's so very  demoralizing that *no matter where* I look in shim, I run into errors. `fallback.c` is full of issues (#382), and I've just briefly looked at `lib/simple_file.c`, and immediately spotted a leak in `simple_file_open_by_handle()` (namely that of `root`). There's basically no healthy code to build fixes upon; one doesn't even know where to start contributing patches (only for them to be ignored nearly forever). Oh well.

Reply:
Thanks for your detailed review comments. I ended up closing this PR as discussed below. The new PR does not call `ReallocatePool` anymore.

> It's so very demoralizing that no matter where I look in shim, I run into errors.

Well, we just have to fix things one step at a time :) I know it can be frustrating to find errors, but taking a step back I don't think it's fair to stay there's no healthy code to build fixes on. Just because there are additional errors that need to be fixed doesn't mean it's not worth it to go ahead and fix some bugs.

Reply:
Thanks for the comment. In general I agree with that attitude. It's just that having to fix (say) three bugs as a dependency chain for the one bug you actually want to fix, is frustrating enough in itself. On top, not seeing any maintainer feedback for a good while in the discussion phase, does not evoke a lot of trust that your work, which has just grown unexpectedly in scope, will not be in vain. "One step at a time" is definitely the way to go, but once it grows into "one series at a time", maintainer overload becomes an even worse obstacle, and a strong discouragement. It makes me doubt my effort is worthwhile. But yes, in general you are right.